Compress layers on push to a v2 registry

dmcgowan opened the pull request:
When buffering to file add support for compressing the tar contents. Since digest should be computed while writing buffer, include digest creation during buffer.

Pull auto detects compression requiring no additional changes.

Ping @jlhawn

jlhawn commented:
LGTM

@jfrazelle Thanks for adding this to the 1.6.0 milestone! It is going to be very important for the release as it significantly reduces payload sizes which in turn reduces registry storage usage and bandwidth requirements.

Smaller downloads == Faster downloads == Happier users :+1:

jfrazelle commented:
ping @crosbymichael this is enhancements to make things faster if its
alright with you we should consider for 1.6

On Fri, Mar 27, 2015 at 2:03 PM, Josh Hawn <notifications@github.com> wrote:

> LGTM
>
> @jfrazelle <https://github.com/jfrazelle> Thanks for adding this to the
> 1.6.0 milestone! It is going to be very important for the release as it
> significantly reduces payload sizes which in turn reduces registry storage
> usage and bandwidth requirements.
>
> Smaller downloads == Faster downloads == Happier users [image: :+1:]
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/11869#issuecomment-87091021>.
>

crosbymichael commented on graph/graph.go:
```diff
@@ -242,18 +245,30 @@ func (graph *Graph) newTempFile() (*os.File, error) {
+		w = gzip.NewWriter(w)
+	}
+	_, err := io.Copy(w, src)
+	if wc, ok := w.(io.Closer); ok {
```
I'm not sure this works, you need to close the underlying file and reset the hash manually.  you cannot cast the multiwriter here.

crosbymichael replied:
nvm, i see that is going on now

dmcgowan replied:
Yes, sorry it is kind of confusing.  Yes it is only for closing the gzip writer, the multiwriter does not need to be closed.

thaJeztah replied:
If it's confusing; perhaps some docs describing the function to prevent future coders tripping over it?

jlhawn replied:
maybe we can be more explicit with the code: 

```go
func bufferToFile(f *os.File, src io.Reader, compress bool) (int64, digest.Digest, error) {
	var (
		d = digest.NewCanonicalDigester() // Uses sha256.
		w = io.MultiWriter(f, &d)
		gz *gzip.Writer
	)

	if compress {
		gz = gzip.NewWriter(w)
		w = gz
	}

	_, err := io.Copy(w, src)
	if gz != nil {
		gz.Close()
	}
	if err != nil {
		return 0, "", err
	}

	n, err := f.Seek(0, os.SEEK_CUR)
	if err != nil {
		return 0, "", err
	}

	if _, err := f.Seek(0, os.SEEK_SET); err != nil {
		return 0, "", err
	}

	return n, d.Digest(), nil
}
```

dmcgowan replied:
Added a comment as @thaJeztah suggested, should be all that is required

thaJeztah replied:
Thanks!

dmcgowan commented:
Updated to make code easier to read based on @jlhawn's recommendation

crosbymichael commented:
Thanks for the readability changes. I'm getting rusty and cannot read code anymore.

jlhawn commented:
@dmcgowan I was talking to @crosbymichael and @NathanMcCauley IRL about this PR and we agreed that if `bufferToFile()` isn't called from anywhere else in the project that we should just not have a `compressed` argument and just *always* compress.

crosbymichael commented:
@dmcgowan can you remove that bool?

dmcgowan committed:
Compress layers on push to a v2 registry

When buffering to file add support for compressing the tar contents. Since digest should be computed while writing buffer, include digest creation during buffer.

Signed-off-by: Derek McGowan <derek@mcgstyle.net> (github: dmcgowan)

dmcgowan commented:
Updated

jlhawn commented:
L<sup>G<sup>T<sup>M</sup></sup></sup>

crosbymichael commented:
@jlhawn that LGTM is not parse-able

LK4D4 commented:
@dmcgowan Could you rebase against master and fix compilation errors?

crosbymichael commented:
LGTM
